### ISSUE TITLE
hili clip viewer: fork or feature?

### ISSUE THREAD

**Issue by breezykermo** [frnsys/rssrs#7]
I'm interested in taking a subset of rssrs' principles (a TUI in which items are row-wise and filter-able) and making a terminal viewer for [hili](https://github.com/frnsys/hili) clips.

The data structure and range of interactions are both smaller than in the case of RSS items. In particular, there is no notion of read/unread items. But the general structure is the same: each item is atomic, has some text that shows in the TUI and an associated URL (the source of the clip), and has optional tags (which are distinctly item-wise in the case of hili clips, rather than attributed to parent 'feeds').

I wonder whether you would be interested in a PR that abstracts the idea of 'item' in rssrs to make it a generic viewer across both RSS items and hili clips. The general idea would be to have different 'tabs' of item types, where a tab represents a viewing mode for a particular item type. Read/unread filtering would be made optional (so that it could be applied to RSS items but not hili clips), and there'd be some parser associated with each tab to fetch items-- async HTTP in the case of RSS, and either local or from a hili server in the case of hili clips.

This would make rssrs into a more generic viewer for any item-wise feed that is representative of some view on external URLs. This is just an idea at the moment, as I've been looking to develop something slightly more flexible for browsing hili clips than [my current adapted server view](https://research.forensic-architecture.org/hili) (which allows [tag-wise filtering](https://research.forensic-architecture.org/hili?t=urbit); but I've started to need proper text-wise search and compositional filtering).

**Comment by breezykermo** [comment]
Follow-up thought here: maybe it doesn't require a complicated re-architecture. Maybe hili just needs to make clips available as an RSS feed, and rssrs provides an option to disable the read/unread functionality so that you can treat it more as a database for search/filter rather than as a news feed.

**Comment by breezykermo** [comment]
Update: turning hili clips into RSS actually works really well, and makes viewing them more generic as well. My [example feed here](https://research.forensic-architecture.org/test.xml) works really nicely with rssrs, or any other RSS reader.

Hili `tags` become RSS `<category>` elements, and all the other fields naturally fit to those in RSS items.

**Comment by frnsys** [comment]
Adding an RSS feed for hili is a great idea--I'm also working on re-writing how the views work here so as to be a bit more modular, which might make your original proposal more doable, but I agree that just using an RSS feed for hili would be better.

**Comment by breezykermo** [comment]
Tabs for different views would be ideal for my use case. Then I'd put all traditional 'news' in one tab, hili clips in another, plus whatever other innovative uses of RSS in others. 

The terminal mail client [aerc](https://aerc-mail.org/) has tabs, for a visual reference for what I'm talking about:

<img width="828" alt="Screen Shot 2020-12-28 at 4 48 34 PM" src="https://user-images.githubusercontent.com/16495856/103234618-6c619080-4940-11eb-9d89-a383923c5df3.png">

**Comment by frnsys** [comment]
Cool, so in the config file there could be some way of specifying tabs (with one tab being one filter)? E.g.

```
- tabs:
  - tag!hili # general news
  - tag:hili # hili clips
```

**Comment by breezykermo** [comment]
Yes. I guess I was thinking that each tab would correspond to some set of feeds. So feeds.txt would have an added hierarchical layer somehow. i.e.:

```
https://hnrss.org/frontpage | Hacker News |  tech | tab1
https://lachlankermode.com/hili/rss.xml | Clips | | tab2
```
 But your version is more succinct / expressive.

**Comment by breezykermo** [comment]
The difficulty with the notion of a tag for hili clips specifically, is that each individual clip has some set of tags (using `<category/>` in RSS parlance). These are item-wise, rather than feed-wise. 

Does the current notion of ‘tag’ in rssrs correspond to `<category/>` items in individual items? Or are they just local to rssrs— and so there is still the option to group/filter by categories in each tab view?

**Comment by frnsys** [comment]
Yeah I was planning on introducing some syntax for custom filters anyways so it can just also be used for tabs, and also so that filters can be based on item properties in addition to feed properties.

> The difficulty with the notion of a tag for hili clips specifically, is that each individual clip has some set of tags (using <category/> in RSS parlance). These are item-wise, rather than feed-wise.
>
> Does the current notion of ‘tag’ in rssrs correspond to <category/> items in individual items? Or are they just local to rssrs— and so there is still the option to group/filter by categories in each tab view?

Right now "tag" just refers to the tags that are added in `feeds.txt`, but it makes sense to also introduce filtering on categories for items. Maybe something like `cat:foo`

**Comment by breezykermo** [comment]
sounds good to me!
